File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    result = 0
    
    def numDecodings(self, s):
        if s[0] == '0':
            return 0
        
        dp = [0] * (len(s) + 1)
        dp[0] = 1
        dp[1] = 1
        for i in range(2, len(s) + 1):
            if int(s[i-1]) >= 1:
                dp[i] += dp[i-1]
            if int(s[i-2:i]) >= 1 and int(s[i-2:i]) <= 26:
                if s[i-2] != "0":
                    # print(i,s[i-1])
                    dp[i] += dp[i-2]
        self.result = dp[len(s)]
        # print("2101")
        # print(dp)
        return self.result

File: 0091-decode-ways/0091-decode-ways.java
class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        int dp[] = new int[n+1];
        
        dp[0] = 1;
        dp[1] = 1;
        
        if (s.charAt(0) == '0'){
            return 0;
        }
        
        // System.out.println(String.valueOf(s.charAt(0)));
        for(int i=2; i<=n; i++){
            int target = s.charAt(i-1) -'0';
            if (target >=1){
                dp[i] += dp[i-1];
            }
            int age = 10;
            
            String tar = s.substring(i-2, i);
            int target2 = Integer.valueOf(tar);
            
            
            if (target2 >= 10 && target2 <= 26){
                dp[i] += dp[i-2];
            }
        }
        return dp[n];
    }
}